File: KeyStats/CloudSyncClient.swift
import Foundation

final class CloudSyncClient {
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(session: URLSession = .shared) {
        self.session = session
        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .iso8601
        self.encoder = JSONEncoder()
        self.encoder.dateEncodingStrategy = .iso8601
    }

    func register(baseURL: URL, username: String, password: String) async throws -> CloudAuthResponse {
        try await authRequest(baseURL: baseURL, path: "auth/register", username: username, password: password)
    }

    func login(baseURL: URL, username: String, password: String) async throws -> CloudAuthResponse {
        try await authRequest(baseURL: baseURL, path: "auth/login", username: username, password: password)
    }

    func listDevices(baseURL: URL, token: String) async throws -> [CloudDevice] {
        let response: CloudDevicesResponse = try await request(
            baseURL: baseURL,
            path: "devices",
            method: "GET",
            token: token
        )
        return response.devices
    }

    func registerDevice(baseURL: URL, token: String, requestBody: CloudRegisterDeviceRequest) async throws -> CloudDevice {
        try await self.request(
            baseURL: baseURL,
            path: "devices",
            method: "POST",
            token: token,
            body: requestBody
        )
    }

    func upsertStats(baseURL: URL, token: String, requestBody: CloudUpsertStatsRequest) async throws {
        let _: CloudUpsertStatsResponse = try await request(
            baseURL: baseURL,
            path: "sync/stats",
            method: "PUT",
            token: token,
            body: requestBody
        )
    }

    func bulkUpsertStats(baseURL: URL, token: String, requestBody: CloudBulkUpsertStatsRequest) async throws {
        let _: CloudBulkUpsertStatsResponse = try await request(
            baseURL: baseURL,
            path: "sync/stats/bulk",
            method: "POST",
            token: token,
            body: requestBody
        )
    }

    func listStats(baseURL: URL, token: String, from: String?, to: String?, deviceId: String?) async throws -> [CloudStatsRecord] {
        var queryItems: [URLQueryItem] = []
        if let from { queryItems.append(URLQueryItem(name: "from", value: from)) }
        if let to { queryItems.append(URLQueryItem(name: "to", value: to)) }
        if let deviceId { queryItems.append(URLQueryItem(name: "device_id", value: deviceId)) }

        let response: CloudStatsListResponse = try await request(
            baseURL: baseURL,
            path: "sync/stats",
            method: "GET",
            token: token,
            queryItems: queryItems
        )
        return response.records
    }

    private func authRequest(baseURL: URL, path: String, username: String, password: String) async throws -> CloudAuthResponse {
        try await request(
            baseURL: baseURL,
            path: path,
            method: "POST",
            body: CloudAuthRequest(username: username, password: password)
        )
    }

    private func request<Response: Decodable, Body: Encodable>(
        baseURL: URL,
        path: String,
        method: String,
        token: String? = nil,
        queryItems: [URLQueryItem] = [],
        body: Body
    ) async throws -> Response {
        let url = apiURL(baseURL: baseURL, path: path, queryItems: queryItems)
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(body)
        return try await perform(request)
    }

    private func request<Response: Decodable>(
        baseURL: URL,
        path: String,
        method: String,
        token: String? = nil,
        queryItems: [URLQueryItem] = []
    ) async throws -> Response {
        let url = apiURL(baseURL: baseURL, path: path, queryItems: queryItems)
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return try await perform(request)
    }

    private func apiURL(baseURL: URL, path: String, queryItems: [URLQueryItem]) -> URL {
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) ?? URLComponents()
        var basePath = components.path
        if basePath.hasSuffix("/") {
            basePath = String(basePath.dropLast())
        }
        components.path = "\(basePath)/api/v1/\(path)"
        if !queryItems.isEmpty {
            components.queryItems = queryItems
        }
        return components.url ?? baseURL
    }

    private func perform<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw CloudSyncError.networkError(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else {
            throw CloudSyncError.invalidResponse
        }

        if (200...299).contains(http.statusCode) {
            if data.isEmpty {
                if let value = "{}".data(using: .utf8) {
                    if let decoded = try? decoder.decode(Response.self, from: value) {
                        return decoded
                    }
                }
            }
            do {
                return try decoder.decode(Response.self, from: data)
            } catch {
                #if DEBUG
                NSLog("[CloudSyncClient] decode failed for %@: %@", request.url?.absoluteString ?? "?", error.localizedDescription)
                #endif
                throw CloudSyncError.invalidResponse
            }
        }

        if let apiError = try? decoder.decode(CloudAPIErrorResponse.self, from: data) {
            throw CloudSyncError.serverError(apiError.error)
        }
        throw CloudSyncError.serverError("HTTP \(http.statusCode)")
    }
}

File: KeyStats.Windows/KeyStats/Services/CloudSyncClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using KeyStats.Models;

namespace KeyStats.Services;

internal sealed class CloudSyncClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true
    };

    private readonly HttpClient _http;

    public CloudSyncClient(HttpClient? httpClient = null)
    {
        _http = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
    }

    public Task<CloudAuthResponse> RegisterAsync(Uri baseUrl, string username, string password, CancellationToken ct = default) =>
        AuthRequestAsync(baseUrl, "auth/register", username, password, ct);

    public Task<CloudAuthResponse> LoginAsync(Uri baseUrl, string username, string password, CancellationToken ct = default) =>
        AuthRequestAsync(baseUrl, "auth/login", username, password, ct);

    public async Task<List<CloudDevice>> ListDevicesAsync(Uri baseUrl, string token, CancellationToken ct = default)
    {
        var response = await RequestAsync<CloudDevicesResponse>(baseUrl, "devices", HttpMethod.Get, token, null, null, ct);
        return response.Devices ?? new List<CloudDevice>();
    }

    public Task<CloudDevice> RegisterDeviceAsync(Uri baseUrl, string token, CloudRegisterDeviceRequest body, CancellationToken ct = default) =>
        RequestAsync<CloudDevice>(baseUrl, "devices", HttpMethod.Post, token, body, null, ct);

    public Task UpsertStatsAsync(Uri baseUrl, string token, CloudUpsertStatsRequest body, CancellationToken ct = default) =>
        RequestAsync<object>(baseUrl, "sync/stats", HttpMethod.Put, token, body, null, ct);

    public Task BulkUpsertStatsAsync(Uri baseUrl, string token, CloudBulkUpsertStatsRequest body, CancellationToken ct = default) =>
        RequestAsync<object>(baseUrl, "sync/stats/bulk", HttpMethod.Post, token, body, null, ct);

    public async Task<List<CloudStatsRecord>> ListStatsAsync(
        Uri baseUrl,
        string token,
        string? from,
        string? to,
        string? deviceId,
        CancellationToken ct = default)
    {
        var query = new List<string>();
        if (!string.IsNullOrWhiteSpace(from)) query.Add($"from={Uri.EscapeDataString(from)}");
        if (!string.IsNullOrWhiteSpace(to)) query.Add($"to={Uri.EscapeDataString(to)}");
        if (!string.IsNullOrWhiteSpace(deviceId)) query.Add($"device_id={Uri.EscapeDataString(deviceId)}");
        var queryString = query.Count > 0 ? "?" + string.Join("&", query) : null;

        var response = await RequestAsync<CloudStatsListResponse>(
            baseUrl, "sync/stats", HttpMethod.Get, token, null, queryString, ct);
        return response.Records ?? new List<CloudStatsRecord>();
    }

    private Task<CloudAuthResponse> AuthRequestAsync(
        Uri baseUrl,
        string path,
        string username,
        string password,
        CancellationToken ct)
    {
        var body = new CloudAuthRequest { Username = username, Password = password };
        return RequestAsync<CloudAuthResponse>(baseUrl, path, HttpMethod.Post, null, body, null, ct);
    }

    private async Task<T> RequestAsync<T>(
        Uri baseUrl,
        string path,
        HttpMethod method,
        string? token,
        object? body,
        string? queryString,
        CancellationToken ct)
    {
        var url = BuildApiUrl(baseUrl, path, queryString);
        using var request = new HttpRequestMessage(method, url);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        if (!string.IsNullOrWhiteSpace(token))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        }

        if (body != null)
        {
            var json = JsonSerializer.Serialize(body, JsonOptions);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            throw new CloudSyncException(ex.Message);
        }

        using (response)
        {
            var data = await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
            if (response.IsSuccessStatusCode)
            {
                if (data.Length == 0)
                {
                    if (typeof(T) == typeof(object))
                    {
                        return default!;
                    }

                    data = Encoding.UTF8.GetBytes("{}");
                }

                try
                {
                    var result = JsonSerializer.Deserialize<T>(data, JsonOptions);
                    if (result == null)
                    {
                        throw new CloudSyncException(Properties.Strings.Sync_Error_InvalidResponse);
                    }

                    return result;
                }
                catch (JsonException)
                {
                    throw new CloudSyncException(Properties.Strings.Sync_Error_InvalidResponse);
                }
            }

            if (data.Length > 0)
            {
                try
                {
                    var apiError = JsonSerializer.Deserialize<CloudAPIErrorResponse>(data, JsonOptions);
                    if (!string.IsNullOrWhiteSpace(apiError?.Error))
                    {
                        throw new CloudSyncException(apiError.Error);
                    }
                }
                catch (JsonException)
                {
                    // fall through
                }
            }

            throw new CloudSyncException($"HTTP {(int)response.StatusCode}");
        }
    }

    private static Uri BuildApiUrl(Uri baseUrl, string path, string? queryString)
    {
        var basePath = baseUrl.AbsolutePath.TrimEnd('/');
        var fullPath = $"{basePath}/api/v1/{path}";
        var builder = new UriBuilder(baseUrl.Scheme, baseUrl.Host, baseUrl.Port, fullPath);
        if (!string.IsNullOrWhiteSpace(queryString))
        {
            builder.Query = queryString.TrimStart('?');
        }

        return builder.Uri;
    }
}
